File: server/app/schemas/document.py
"""Document pipeline DTOs (Prompt 11).

All responses use camelCase JSON via the shared ``APIModel`` alias generator.
Extracted fields are returned with muted display values where sensitive; the
server never returns the private ``file_ref`` or raw stored text.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.common import APIModel

_OcrConfidence = Literal["high", "needs_review", "low"]


class DocumentTypeOut(APIModel):
    """One catalog entry (also powers the accepted-format list for the UI)."""

    code: str
    kind: str
    name_en: str
    localized_names: dict = Field(default_factory=dict)
    ocr_supported: bool = True
    accepted_formats: list[str] = Field(default_factory=list)
    guidance: dict = Field(default_factory=dict)


class ExtractedFieldOut(APIModel):
    """One extracted field with its display-safe masked copy."""

    key: str
    label: str
    value: str
    masked: str | None = None
    reliable: bool | None = None


class UserDocumentOut(APIModel):
    """Public view of an uploaded document (owner may also fetch the file)."""

    id: str
    user_id: str
    scheme_code: str | None = None
    required_name: str | None = None
    file_name: str
    file_extension: str
    file_size_bytes: int
    mime_type: str
    status: str
    ocr_confidence: str | None = None
    detected_type: str | None = None
    detection_confidence: float | None = None
    type_matches: bool | None = None
    detection_code: str | None = None
    extracted_fields: list[ExtractedFieldOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    processed_at: datetime | None = None
    reviewed_at: datetime | None = None


class UserDocumentListOut(APIModel):
    items: list[UserDocumentOut] = Field(default_factory=list)
    total: int = 0


class OcrSubmitIn(APIModel):
    """OCR text pushed from the browser OCR adapter (tesseract)."""

    text: str = Field(min_length=1)


class OcrResultOut(APIModel):
    """Verdict of the OCR + type-detection pass."""

    document_id: str
    detected_type: str | None = None
    type_name: str = ""
    confidence: float = 0.0
    ocr_confidence: _OcrConfidence = "needs_review"
    detection_code: Literal["matched", "partial", "unrecognised", "mismatch"] = "unrecognised"
    extracted_fields: list[ExtractedFieldOut] = Field(default_factory=list)
    needs_manual_selection: bool = True


class DocumentTypeConfirmIn(APIModel):
    document_type: str


class DocumentReviewIn(APIModel):
    """User-confirmed field corrections (never silently applied)."""

    fields: list[ExtractedFieldOut] = Field(default_factory=list)
    note: str | None = None


class DocumentReplaceIn(APIModel):
    """Optional replacement payload for a document (client sends a new file)."""

    # Present for symmetry; upload routes use multipart bodies.
    pass


class ChecklistItemOut(APIModel):
    required: dict = Field(default_factory=dict)
    status: str
    user_document: UserDocumentOut | None = None
    is_missing: bool = True
    guidance: str | None = None
    official_source_url: str | None = None


class DocumentReadinessOut(APIModel):
    """Per-scheme pre-check summary. Never an official approval score."""

    scheme_code: str
    required_count: int = 0
    uploaded_count: int = 0
    missing_count: int = 0
    needs_review_count: int = 0
    percent: int = 0
    items: list[ChecklistItemOut] = Field(default_factory=list)
    disclaimer: str = ""


class OcrUnavailableOut(APIModel):
    ok: bool = False
    message: str = ""


class UploadOut(APIModel):
    """Multipart upload envelope."""

    document: UserDocumentOut
    ocr_available: bool


def _extracted_to_out(fields: list[dict]) -> list[ExtractedFieldOut]:
    out = []
    for field in fields:
        if not isinstance(field, dict):
            continue
        out.append(
            ExtractedFieldOut(
                key=str(field.get("key", "")),
                label=str(field.get("label", "")),
                value=str(field.get("value", "")),
                masked=field.get("masked") if field.get("masked") else None,
                reliable=bool(field.get("reliable")),
            )
        )
    return out


__all__ = [
    "ChecklistItemOut",
    "DocumentReadinessOut",
    "DocumentReplaceIn",
    "DocumentReviewIn",
    "DocumentTypeConfirmIn",
    "DocumentTypeOut",
    "ExtractedFieldOut",
    "OcrResultOut",
    "OcrUnavailableOut",
    "UploadOut",
    "UserDocumentListOut",
    "UserDocumentOut",
    "_extracted_to_out",
]

File: shared/src/domain/document.ts
import type { DocumentKind, RequiredDocument } from "./scheme";
import type { ISODateString, LanguageCode, UUID } from "./common";

/**
 * Canonical codes used by the document catalog (`document_types`) and the OCR
 * detector. Scheme requirements reference these by `name`; OCR maps the
 * recognized text to one of these codes.
 */
export type DocumentCode =
  | "AADHAAR"
  | "PAN_CARD"
  | "RATION_CARD"
  | "BANK_PASSBOOK"
  | "BANK_ACCOUNT"
  | "INCOME_CERTIFICATE"
  | "COMMUNITY_CERTIFICATE"
  | "CASTE_CERTIFICATE"
  | "RESIDENCE_CERTIFICATE"
  | "DISABILITY_CERTIFICATE"
  | "BIRTH_CERTIFICATE"
  | "MARK_SHEET"
  | "VOTER_ID"
  | "PASSPORT"
  | "LAND_RECORD"
  | "PHOTOGRAPH"
  | "MARRIAGE_CERTIFICATE"
  | "APPLICATION_FORM"
  | "OTHER";

/**
 * Lifecycle of a user's document through upload → OCR → review.
 * `missing` is a derived state only (no DB row exists yet).
 */
export type DocumentStatus =
  | "missing"
  | "uploaded"
  | "processing"
  | "processed"
  | "needs_review"
  | "matches"
  | "mismatch"
  | "unsupported"
  | "ocr_failed"
  | "user_confirmed";

/** Simplified OCR quality indicator — never proof of genuineness. */
export type OcrConfidence = "high" | "needs_review" | "low";

/** One extracted field with a masked display form for sensitive values. */
export interface ExtractedField {
  key: string;
  label: string;
  value: string;
  /** Masked copy safe to show the user (e.g. "XXXX XXXX 1234"). */
  masked?: string;
  /** True when the detector was reasonably sure of this field. */
  reliable?: boolean;
}

/** Catalog entry for one recognisable document type (server `document_types`). */
export interface DocumentTypeInfo {
  code: DocumentCode;
  kind: DocumentKind;
  nameEn: string;
  localizedNames: Partial<Record<LanguageCode, string>>;
  ocrSupported: boolean;
  acceptedFormats: string[];
  /** View box for the "How to obtain it" guidance links. */
  guidance?: DocumentGuidance;
}

/** Guidance copy shown for a missing document type. */
export interface DocumentGuidance {
  summaryEn: string;
  officialSourceUrl?: string;
}

/** A citizen's uploaded document (server `user_documents`, + extraction/review). */
export interface UserDocument {
  id: UUID;
  userId: UUID;
  schemeCode?: string;
  /** The scheme requirement this upload targets (RequiredDocument.name). */
  requiredName?: string;
  fileName: string;
  fileExtension: string;
  fileSizeBytes: number;
  mimeType: string;
  status: DocumentStatus;
  ocrConfidence: OcrConfidence | null;
  detectedType: DocumentCode | null;
  detectionConfidence: number | null;
  /**
   * True when the detected type matches the scheme's required document.
   * `null` when there is nothing to match against yet.
   */
  typeMatches: boolean | null;
  /** Detected/extracted fields after OCR + user correction. */
  extractedFields: ExtractedField[];
  required?: RequiredDocument;
  createdAt: ISODateString;
  updatedAt: ISODateString;
  processedAt?: ISODateString;
  reviewedAt?: ISODateString;
}

/** Result of the OCR + document-type-detection pass for one upload. */
export interface OcrResult {
  documentId: UUID;
  detectedType: DocumentCode | null;
  typeName: string;
  confidence: number;
  ocrConfidence: OcrConfidence;
  /** Machine-readable reason for the verdict ("matched"/"unrecognised"/...). */
  detectionCode: "matched" | "partial" | "unrecognised" | "mismatch";
  extractedFields: ExtractedField[];
  /** True so the UI can hand off to manual type selection. */
  needsManualSelection: boolean;
}

/** Manual confirmation/correction of the document type by the user. */
export interface DocumentTypeConfirmRequest {
  documentType: DocumentCode;
}

/** Review submission: user confirms or corrects the extracted information. */
export interface DocumentReviewRequest {
  /** Corrected field values the user verified (all fields incl. corrections). */
  fields?: ExtractedField[];
}

/** One row in a scheme's document checklist. */
export interface ChecklistItem {
  required: RequiredDocument;
  status: DocumentStatus;
  /** Already-uploaded document for this requirement (when one exists). */
  userDocument?: UserDocument | null;
  /** True when this claim still needs a document to be "complete". */
  isMissing: boolean;
  /** Citizen-friendly guidance when missing / needs attention. */
  guidance?: string;
  officialSourceUrl?: string;
}

/** Per-scheme readiness summary (CiviServe pre-check only). */
export interface DocumentReadiness {
  schemeCode: string;
  requiredCount: number;
  uploadedCount: number;
  missingCount: number;
  needsReviewCount: number;
  /** 0..100 — NOT an official approval score. */
  percent: number;
  items: ChecklistItem[];
  /**
   * Fixed disclaimer copy: pre-check only, final acceptance is decided by the
   * relevant government authority.
   */
  disclaimer: string;
}

export interface UserDocumentListResponse {
  items: UserDocument[];
  total: number;
}